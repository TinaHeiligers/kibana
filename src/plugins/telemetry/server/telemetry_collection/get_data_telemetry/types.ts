/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { DataTelemetryType, DataPatternName } from './constants';

export interface DataTelemetryBasePayload {
  index_count: number;
  ecs_index_count?: number;
  doc_count?: number;
  size_in_bytes?: number;
}

export interface DataTelemetryDocument extends DataTelemetryBasePayload {
  data_stream?: {
    dataset?: string;
    type?: DataTelemetryType | string; // The union of types is to help autocompletion with some known `data_stream.type`s
  };
  package?: {
    name: string;
  };
  shipper?: string;
  pattern_name?: DataPatternName;
}

export type DataTelemetryPayload = DataTelemetryDocument[];

export interface DataTelemetryIndex {
  name: string;
  packageName?: string; // Populated by Ingest Manager at `_meta.package.name`
  managedBy?: string; // Populated by Ingest Manager at `_meta.managed_by`
  dataStreamDataset?: string; // To be obtained from `mappings.data_stream.dataset` if it's a constant keyword
  dataStreamType?: string; // To be obtained from `mappings.data_stream.type` if it's a constant keyword
  shipper?: string; // To be obtained from `_meta.beat` if it's set
  isECS?: boolean; // Optional because it can't be obtained via Monitoring.

  // The fields below are optional because we might not be able to obtain them if the user does not
  // have access to the index.
  docCount?: number;
  sizeInBytes?: number;
}

export type AtLeastOne<T, U = { [K in keyof T]: Pick<T, K> }> = Partial<T> & U[keyof U];

export type DataDescriptor = AtLeastOne<{
  packageName: string;
  dataStreamDataset: string;
  dataStreamType: string;
  shipper: string;
  patternName: DataPatternName; // When found from the list of the index patterns
}>;
export interface IndexStats {
  indices: {
    [indexName: string]: {
      total: {
        docs: {
          count: number;
          deleted: number;
        };
        store: {
          size_in_bytes: number;
        };
      };
    };
  };
}

export interface IndexMappings {
  [indexName: string]: {
    mappings: {
      _meta?: {
        beat?: string;

        // Ingest Manager provided metadata
        package?: {
          name?: string;
        };
        managed_by?: string; // Typically "ingest-manager"
      };
      properties: {
        data_stream?: {
          properties: {
            dataset?: {
              type: string;
              value?: string;
            };
            type?: {
              type: string;
              value?: string;
            };
          };
        };
        ecs?: {
          properties: {
            version?: {
              type: string;
            };
          };
        };
      };
    };
  };
}
