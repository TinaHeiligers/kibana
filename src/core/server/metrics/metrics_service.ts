/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import numeral from '@elastic/numeral';
import { ReplaySubject } from 'rxjs';
import { first } from 'rxjs/operators';
import { CoreService } from '../../types';
import { CoreContext } from '../core_context';
import { Logger } from '../logging';
import { InternalHttpServiceSetup } from '../http';
import { InternalMetricsServiceSetup, InternalMetricsServiceStart, OpsMetrics } from './types';
import { OpsMetricsCollector } from './ops_metrics_collector';
import { opsConfig, OpsConfigType } from './ops_config';

interface MetricsServiceSetupDeps {
  http: InternalHttpServiceSetup;
}

/** @internal */
export class MetricsService
  implements CoreService<InternalMetricsServiceSetup, InternalMetricsServiceStart> {
  private readonly logger: Logger;
  private metricsCollector?: OpsMetricsCollector;
  private collectInterval?: NodeJS.Timeout;
  private metrics$ = new ReplaySubject<OpsMetrics>(1);
  private service?: InternalMetricsServiceSetup;

  constructor(private readonly coreContext: CoreContext) {
    this.logger = coreContext.logger.get('metrics');
  }

  public async setup({ http }: MetricsServiceSetupDeps): Promise<InternalMetricsServiceSetup> {
    const config = await this.coreContext.configService
      .atPath<OpsConfigType>(opsConfig.path)
      .pipe(first())
      .toPromise();

    this.metricsCollector = new OpsMetricsCollector(http.server, {
      logger: this.logger,
      ...config.cGroupOverrides,
    });

    await this.refreshMetrics();

    this.collectInterval = setInterval(() => {
      this.refreshMetrics(); // I'm logging the ops metrics in here
    }, config.interval.asMilliseconds());

    const metricsObservable = this.metrics$.asObservable();

    this.service = {
      collectionInterval: config.interval.asMilliseconds(),
      getOpsMetrics$: () => metricsObservable,
    };

    return this.service;
  }

  public async start(): Promise<InternalMetricsServiceStart> {
    if (!this.service) {
      throw new Error('#setup() needs to be run first');
    }
    return this.service;
  }

  private extractOpsLogsData({ process, os }: Partial<OpsMetrics>): string {
    const memoryLogEntry = process?.memory?.heap?.used_in_bytes ?? 0;
    const uptimeLogEntry = process?.uptime_in_millis ?? 0;
    const loadLogEntry = os?.load ?? [];
    const delayLogEntry = process?.event_loop_delay ?? 0;
    const memoryLogEntryInMB = numeral(memoryLogEntry).format('0.0b');
    // legacy logging has uptime in seconds (from Hapi Good), whereas we have that in milliseconds in the metrics service
    const uptimeLogEntryPretty = numeral(uptimeLogEntry / 1000).format('00:00:00');
    const loadLogEntryPretty = [...Object.values(loadLogEntry)]
      .map((val: number) => {
        return numeral(val).format('0.00');
      })
      .join(' ');
    const delayLogEntryPretty = numeral(delayLogEntry).format('0.000');

    return `memory: ${memoryLogEntryInMB} uptime: ${uptimeLogEntryPretty} load: [${loadLogEntryPretty}] delay: ${delayLogEntryPretty}`;
  }

  private async refreshMetrics() {
    this.logger.debug('Refreshing metrics');
    const metrics = await this.metricsCollector!.collect();
    // TODO: decide if this is the correct place to log the ops metrics from
    const opsMetricsPretty = this.extractOpsLogsData(metrics);
    this.logger.info(opsMetricsPretty);

    this.metricsCollector!.reset();
    this.metrics$.next(metrics);
  }

  public async stop() {
    if (this.collectInterval) {
      clearInterval(this.collectInterval);
    }
    this.metrics$.complete();
  }
}
