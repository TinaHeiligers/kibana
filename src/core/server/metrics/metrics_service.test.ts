/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import moment from 'moment';

import { configServiceMock } from '../config/mocks';
import { mockOpsCollector } from './metrics_service.test.mocks';
import { MetricsService } from './metrics_service';
import { mockCoreContext } from '../core_context.mock';
import { httpServiceMock } from '../http/http_service.mock';
import { loggingSystemMock } from '../logging/logging_system.mock';
import { analyticsServiceMock } from '../mocks';
import { firstValueFrom, Observable } from 'rxjs';
import { OpsMetrics } from './types';

const testInterval = 100;

const dummyMetrics = { metricA: 'value', metricB: 'otherValue' };

const logger = loggingSystemMock.create();

const createNextEmission = async (getOpsMetrics$: () => Observable<OpsMetrics>) => {
  jest.advanceTimersByTime(testInterval);
  return firstValueFrom(getOpsMetrics$());
};

const createOpsMetrics = (testMetrics: any) => {
  mockOpsCollector.collect.mockResolvedValueOnce(testMetrics);
};

describe('MetricsService', () => {
  const httpMock = httpServiceMock.createInternalSetupContract();
  const analyticsMock = analyticsServiceMock.createAnalyticsServiceSetup();
  let metricsService: MetricsService;

  beforeEach(() => {
    jest.useFakeTimers();

    const configService = configServiceMock.create({
      atPath: { interval: moment.duration(testInterval) },
    });
    const coreContext = mockCoreContext.create({ logger, configService });
    metricsService = new MetricsService(coreContext);
  });

  afterEach(() => {
    jest.clearAllMocks();
    jest.clearAllTimers();
  });

  describe('#start', () => {
    it('invokes setInterval with the configured interval', async () => {
      await metricsService.setup({ http: httpMock, analytics: analyticsMock });
      await metricsService.start();

      expect(setInterval).toHaveBeenCalledTimes(1);
      expect(setInterval).toHaveBeenCalledWith(expect.any(Function), testInterval);
    });

    it('collects the metrics at every interval', async () => {
      mockOpsCollector.collect.mockResolvedValue(dummyMetrics);

      await metricsService.setup({ http: httpMock, analytics: analyticsMock });
      await metricsService.start();

      expect(mockOpsCollector.collect).toHaveBeenCalledTimes(1);

      jest.advanceTimersByTime(testInterval);
      expect(mockOpsCollector.collect).toHaveBeenCalledTimes(2);

      jest.advanceTimersByTime(testInterval);
      expect(mockOpsCollector.collect).toHaveBeenCalledTimes(3);
    });

    it('resets the collector after each collection', async () => {
      createOpsMetrics(dummyMetrics);

      await metricsService.setup({ http: httpMock, analytics: analyticsMock });
      const { getOpsMetrics$ } = await metricsService.start();

      expect(mockOpsCollector.collect).toHaveBeenCalledTimes(1);
      expect(mockOpsCollector.reset).toHaveBeenCalledTimes(1);

      await createNextEmission(getOpsMetrics$);
      expect(mockOpsCollector.collect).toHaveBeenCalledTimes(2);
      expect(mockOpsCollector.reset).toHaveBeenCalledTimes(2);

      await createNextEmission(getOpsMetrics$);
      expect(mockOpsCollector.collect).toHaveBeenCalledTimes(3);
      expect(mockOpsCollector.reset).toHaveBeenCalledTimes(3);
    });

    it('throws when called before setup', async () => {
      await expect(metricsService.start()).rejects.toThrowErrorMatchingInlineSnapshot(
        `"#setup() needs to be run first"`
      );
    });

    it('emits the last value on each getOpsMetrics$ call', async () => {
      const firstMetrics = { metric: 'first' };
      const secondMetrics = { metric: 'second' };
      createOpsMetrics(firstMetrics);
      createOpsMetrics(secondMetrics);

      await metricsService.setup({ http: httpMock, analytics: analyticsMock });
      const { getOpsMetrics$ } = await metricsService.start();

      expect(await createNextEmission(getOpsMetrics$)).toEqual({ metric: 'first' });
      expect(await createNextEmission(getOpsMetrics$)).toEqual({ metric: 'second' });
    });

    it('logs the metrics at every interval', async () => {
      const firstMetrics = {
        process: {
          memory: { heap: { used_in_bytes: 100 } },
          uptime_in_millis: 1500,
          event_loop_delay: 50,
        },
        os: {
          load: {
            '1m': 10,
            '5m': 20,
            '15m': 30,
          },
        },
      };
      const secondMetrics = {
        process: {
          memory: { heap: { used_in_bytes: 200 } },
          uptime_in_millis: 3000,
          event_loop_delay: 100,
        },
        os: {
          load: {
            '1m': 20,
            '5m': 30,
            '15m': 40,
          },
        },
      };

      const opsLogger = logger.get('metrics', 'ops');

      createOpsMetrics(firstMetrics);
      createOpsMetrics(secondMetrics);

      await metricsService.setup({ http: httpMock, analytics: analyticsMock });
      const { getOpsMetrics$ } = await metricsService.start();

      await createNextEmission(getOpsMetrics$);
      const opsLogs = loggingSystemMock.collect(opsLogger).debug;
      expect(opsLogs.length).toEqual(2);
      expect(opsLogs[0][1]).not.toEqual(opsLogs[1][1]);
    });

    it('omits metrics from log message if they are missing or malformed', async () => {
      const opsLogger = logger.get('metrics', 'ops');
      createOpsMetrics({ secondMetrics: 'metrics' });
      await metricsService.setup({ http: httpMock, analytics: analyticsMock });
      await metricsService.start();
      expect(loggingSystemMock.collect(opsLogger).debug[0]).toMatchInlineSnapshot(`
        Array [
          "",
          Object {
            "event": Object {
              "category": Array [
                "process",
                "host",
              ],
              "kind": "metric",
              "type": Array [
                "info",
              ],
            },
            "host": Object {
              "os": Object {
                "load": Object {
                  "15m": undefined,
                  "1m": undefined,
                  "5m": undefined,
                },
              },
            },
            "process": Object {
              "eventLoopDelay": undefined,
              "eventLoopDelayHistogram": undefined,
              "memory": Object {
                "heap": Object {
                  "usedInBytes": undefined,
                },
              },
              "uptime": undefined,
            },
          },
        ]
      `);
    });

    it('reports ops metrics as an event to analytics at every interval', async () => {
      const firstMetrics = {
        process: {
          memory: { heap: { used_in_bytes: 100 } },
          uptime_in_millis: 1500,
          event_loop_delay: 50,
        },
        os: {
          load: {
            '1m': 10,
            '5m': 20,
            '15m': 30,
          },
        },
      };
      const secondMetrics = {
        process: {
          memory: { heap: { used_in_bytes: 200 } },
          uptime_in_millis: 3000,
          event_loop_delay: 100,
        },
        os: {
          load: {
            '1m': 20,
            '5m': 30,
            '15m': 40,
          },
        },
      };

      createOpsMetrics(firstMetrics);
      createOpsMetrics(secondMetrics);

      await metricsService.setup({ http: httpMock, analytics: analyticsMock });

      const { getOpsMetrics$ } = await metricsService.start();

      expect(analyticsMock.registerEventType).toHaveBeenCalledTimes(1);
      expect(analyticsMock.reportEvent).toHaveBeenCalledTimes(1);

      await createNextEmission(getOpsMetrics$);

      expect(analyticsMock.registerEventType).toHaveBeenCalledTimes(1);
      expect(analyticsMock.reportEvent).toHaveBeenCalledTimes(2);
    });

    it('logs reporting ops metrics as an event to analytics at every interval', async () => {
      const firstMetrics = {
        process: {
          memory: { heap: { used_in_bytes: 100 } },
          uptime_in_millis: 1500,
          event_loop_delay: 50,
        },
        os: {
          load: {
            '1m': 10,
            '5m': 20,
            '15m': 30,
          },
        },
      };
      const secondMetrics = {
        process: {
          memory: { heap: { used_in_bytes: 200 } },
          uptime_in_millis: 3000,
          event_loop_delay: 100,
        },
        os: {
          load: {
            '1m': 20,
            '5m': 30,
            '15m': 40,
          },
        },
      };

      const analyticsMetricsLogger = logger.get('metrics', 'metrics_analytics');
      createOpsMetrics(firstMetrics);
      createOpsMetrics(secondMetrics);

      await metricsService.setup({ http: httpMock, analytics: analyticsMock });
      const { getOpsMetrics$ } = await metricsService.start();

      await createNextEmission(getOpsMetrics$);

      const analyticsMetricsLogs = loggingSystemMock.collect(analyticsMetricsLogger).info;
      expect(analyticsMetricsLogs.length).toEqual(2);
      expect(analyticsMetricsLogs[0]).not.toEqual(analyticsMetricsLogs[1]);
    });
  });

  describe('#stop', () => {
    it('stops the metrics interval', async () => {
      await metricsService.setup({ http: httpMock, analytics: analyticsMock });
      const { getOpsMetrics$ } = await metricsService.start();

      expect(mockOpsCollector.collect).toHaveBeenCalledTimes(1);

      jest.advanceTimersByTime(testInterval);
      expect(mockOpsCollector.collect).toHaveBeenCalledTimes(2);

      await metricsService.stop();
      jest.advanceTimersByTime(10 * testInterval);
      expect(mockOpsCollector.collect).toHaveBeenCalledTimes(2);

      getOpsMetrics$().subscribe({ complete: () => {} });
    });

    it('completes the metrics observable', async () => {
      await metricsService.setup({ http: httpMock, analytics: analyticsMock });
      const { getOpsMetrics$ } = await metricsService.start();

      let completed = false;

      getOpsMetrics$().subscribe({
        complete: () => {
          completed = true;
        },
      });

      await metricsService.stop();

      expect(completed).toEqual(true);
    });
  });
});
