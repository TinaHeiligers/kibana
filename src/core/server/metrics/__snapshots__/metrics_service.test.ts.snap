// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MetricsService #start logs the metrics at every interval 1`] = `
Array [
  Array [
    "memory: 100.0B uptime: 0:00:02 load: [10.00 20.00 30.00] delay: 50.000",
  ],
]
`;

exports[`MetricsService #start logs the metrics at every interval 2`] = `
Array [
  Array [
    "memory: 100.0B uptime: 0:00:02 load: [10.00 20.00 30.00] delay: 50.000",
  ],
  Array [
    "memory: 200.0B uptime: 0:00:03 load: [20.00 30.00 40.00] delay: 100.000",
  ],
]
`;
